h2/test_h2_headers.py

```diff
@@ -106,3 +107,91 @@ def test_random_header(self):
         self.start_tempesta()
         self.start_all_clients()
         self.wait_while_busy(curl)
+
+class CurlTest(tester.TempestaTest):
+
+    clients = [
+        {
+            'id' : 'curl',
+            'type' : 'external',
+            'binary' : 'curl',
+            'cmd_args' : (
+                '-k '
+                'https://${tempesta_ip}/ '
+                )
+        },
+    ]
+
+    def test(self):
+        curl = self.get_client('curl')
+
+        self.start_all_servers()
+        self.start_tempesta()
+        self.start_all_clients()
+        self.wait_while_busy(curl)
+
+        self.assertEqual(0, curl.returncode,
+                         msg=("Curl return code is not 0 (%d)." %
+                              (curl.returncode)))
+
+class TransformHeadersInLC(CurlTest):
+    '''https://tools.ietf.org/html/rfc7540#section-8.1.2
+    Just as in HTTP/1.x, header field names are strings of ASCII
+    characters that are compared in a case-insensitive fashion.  However,
+    header field names MUST be converted to lowercase prior to their
+    encoding in HTTP/2.  A request or response containing uppercase
+    header field names MUST be treated as malformed.
+    Ask curl to make an h2 request, if header field names are not converted to
+    lowercase then curl return code is not 0.
+    '''
+
+    backends = [
+        {
+            'id' : 'nginx',
+            'type' : 'nginx',
+            'port' : '8000',
+            'status_uri' : 'http://${server_ip}:8000/nginx_status',
+            'config' : NGINX_CONFIG % """
+add_header X-Extra-Data1 "q";
+add_header X-Extra-Data2 "q";
+add_header X-Extra-Data1 "q";
+
+return 200;
+""",
+        }
+    ]
+
+    tempesta = {
+        'config' : TEMPESTA_CONFIG % "",
+    }
+
+class TransformHeadersInLCCache(CurlTest):
+    '''https://tools.ietf.org/html/rfc7540#section-8.1.2
+    Just as in HTTP/1.x, header field names are strings of ASCII
+    characters that are compared in a case-insensitive fashion.  However,
+    header field names MUST be converted to lowercase prior to their
+    encoding in HTTP/2.  A request or response containing uppercase
+    header field names MUST be treated as malformed.
+    Ask curl to make an h2 request, if header field names are not converted to
+    lowercase then curl return code is not 0.
+    '''
+
+    backends = [
+        {
+            'id' : 'nginx',
+            'type' : 'nginx',
+            'port' : '8000',
+            'status_uri' : 'http://${server_ip}:8000/nginx_status',
+            'config' : NGINX_CONFIG % """
+add_header X-Extra-Data1 "q";
+add_header X-Extra-Data2 "q";
+add_header X-Extra-Data1 "q";
+
+return 200;
+""",
+        }
+    ]
+
+    tempesta = {
+        'config' : TEMPESTA_CONFIG % "cache_fulfill * *;",
```

Comment:
The Test works incorrect: in cached mode minimum 2 requests are required: one to populate the cache and one to use it. Also need to check performance counters, this allows to avoid situations when a response can't be cached due to bug and the second response is also comes from the origin server.

Reply:
The comment _was_ addressed.